### DISCUSSION TITLE
dnsmasq via infoblox and dhcp options

### DISCUSSION THREAD

**Discussion by HappyRizi** [ltsp/ltsp#529] [Q&A]
Many thanks to people who involved in developing the LTSP solution.
I have been using ltsp since 2009 without any problems.

Server: ubuntu 20.04.3 LTS with one ethernet.
LTSP: ltsp 21.01-1~ubuntu20.04.1

I upgraded our ltsp5 to ltsp20. The old version worked fin with external dhcp server (infoblox). But now the client does not receive any boot file from ltsp server on the new one.
There are no documentation regarding parameters which involving external dhcp server.

I am using the dhcp options as following:

```
dhcp-option=17	"/srv/tftp/ltsp"
dhcp-option=66	"ltsp server IP-address"
dhcp-option=67	"/ltsp/ltsp.ipxe"

Clients gets IP address from infoblox and nothing happend:
-------------------(client boot)----------------------
CLIENT MAC ADDR: 2C 27 D7 ZD 12 E7 GUID: 00A07B09 4676 E011 0000 2C27DZD12E7
CLIENT IP: 10.WWW.XXX.YYY MASK: 255.255.255.0 DHCP IP: 10.220.ZZZ.PPP
GATEWAY IP: 10.WWW.XXX.1
_
-------(Stops here and nothing happens)--------------------
---------------(on the ltsp server)-----------------------------------
journalctl -fu dnsmasq
sysuser@ltspserver:~$ journalctl -fu dnsmasq
-- Logs begin at Thu 2021-08-19 15:44:24 CEST. --
Sep 06 09:42:10 ltsp1 dnsmasq-dhcp[49095]: PXE(enp10s0) 00:24:81:e3:af:84 proxy
Sep 10 08:24:00 ltsp1 dnsmasq-tftp[49095]: error 0 TFTP Aborted received from 10.www.xxx.211
Sep 10 08:24:00 ltsp1 dnsmasq-tftp[49095]: failed sending /srv/tftp/ltsp/ltsp.ipxe to 10.www.xxx.211    
Sep 10 08:24:00 ltsp1 dnsmasq-tftp[49095]: sent /srv/tftp/ltsp/ltsp.ipxe to 10.www.xxx.211			
---------------------------------------
```

I think there are something wrong with the infoblox parameters. Is there someone who tried this with infoblox and please let me know if my dhcp options 17, 66 and 67 are correct.
Thanks.

**Comment by alkisg** [comment]
Hi HappyRizi,

The easiest way is to use infoblox as the normal DHCP server that gives IP leases and not boot information (i.e. to delete all the dhcp-options that you mentioned),
and then to run `ltsp dnsmasq` on the ltsp server, so that it's configured in proxy DHCP mode.

Try that first.
If you don't like that method, then you can use the dnsmasq template of LTSP, you may find it in /usr/share/ltsp/server/dnsmasq/ltsp-dnsmasq.conf. You may see the necessary dhcp-options and pxe-services from there. It's not just 2-3 lines, there are many needed, to serve BIOS and UEFI and iPXE and Raspberry Pi clients.

Btw, if you want to paste configuration settings etc, put them inside triple backticks: \`\`\`
I updated your post with them so that it displays better.

**Reply by HappyRizi** [reply]
Thank you for your reply, alkisg
The ltsp server and the clients are located on different VLAN. INFOBLOX make sure that clients get ip-addresses. I am using the first method :  "using infoblox as the normal DHCP server that gives IP leases ..." 
All dhcp options are removed from infoblox, I have only BOOTP Settings:
Boot File: ltsp.ipxe
Next Server: ltsp-server-ip
Boot Server: ltsp-server-ip
Or should I use "Deny BOOTP Requests" ?

In my /etc/dnsmasq.d/ltsp-dsnmasq.conf I have:
port=0
enable-tftp
tftp-root=/srv/tftp

When client boot, I get 
Sep 15 11:24:01 ltsp1 dnsmasq-tftp[225476]: TFTP root is /srv/tftp
Sep 15 11:24:01 ltsp1 systemd[1]: Started dnsmasq - A lightweight DHCP and caching DNS server.
Sep 15 12:01:16 ltsp1 dnsmasq-tftp[225476]: file /srv/tftp/ltsp.ipxe not found
Sep 15 12:02:29 ltsp1 dnsmasq-tftp[225476]: file /srv/tftp/ltsp.ipxe not found

when I change tftp-root=/srv/tftp to tftp-root=/srv/tftp/ltsp in ltsp-dnsmasq.conf I get : 
Sep 15 13:01:36 ltsp1 dnsmasq-tftp[226107]: error 0 TFTP Aborted received from 10.112.16.211
Sep 15 13:01:36 ltsp1 dnsmasq-tftp[226107]: failed sending /srv/tftp/ltsp/ltsp.ipxe to 10.112.16.211
Sep 15 13:01:36 ltsp1 dnsmasq-tftp[226107]: sent /srv/tftp/ltsp/ltsp.ipxe to 10.112.16.211
Sep 15 13:03:28 ltsp1 dnsmasq-tftp[226107]: error 0 TFTP Aborted received from 10.112.16.192
Sep 15 13:03:28 ltsp1 dnsmasq-tftp[226107]: failed sending /srv/tftp/ltsp/ltsp.ipxe to 10.112.16.192
Sep 15 13:03:28 ltsp1 dnsmasq-tftp[226107]: sent /srv/tftp/ltsp/ltsp.ipxe to 10.112.16.192

Can you help please?

**Comment by alkisg** [comment]
If it's on a different VLAN, the first method won't work without a DHCP relay.

Use the second method; read the ltsp-dnsmasq.conf template and put all the related lines into your infoblox.
A plain `Boot File` setting is not enough.

I.e. all lines from 35 and below in this file:
https://github.com/ltsp/ltsp/blob/main/ltsp/server/dnsmasq/ltsp-dnsmasq.conf

**Comment by HappyRizi** [comment]
But my /etc/dnsmasq.d/ltsp-dnsmasq.conf file already contain all 20+ related lines:
port=0
enable-tftp
tftp-root=/srv/tftp
dhcp-range=10.112.16.10,10.112.16.250,12h
dhcp-range=set:proxy,10.112.16.10,proxy,255.255.255.0
dhcp-option=option:dns-server,10.220.2.5
dhcp-match=set:iPXE,175,39
dhcp-match=set:X86PC,option:client-arch,0
dhcp-match=set:X86-64_EFI,option:client-arch,7
dhcp-match=set:X86-64_EFI,option:client-arch,9
dhcp-mac=set:rpi,b8:27:eb:*:*:*
dhcp-mac=set:rpi,dc:a6:32:*:*:*
pxe-service=tag:proxy,tag:!iPXE,X86PC,"undionly.kpxe",ltsp/undionly.kpxe
pxe-service=tag:proxy,tag:!iPXE,X86-64_EFI,"snponly.efi",ltsp/snponly.efi
pxe-service=tag:proxy,tag:iPXE,X86PC,"ltsp.ipxe",ltsp/ltsp.ipxe
pxe-service=tag:proxy,tag:iPXE,X86-64_EFI,"ltsp.ipxe",ltsp/ltsp.ipxe
pxe-service=tag:rpi,X86PC,"Raspberry Pi Boot   ",unused
dhcp-boot=tag:!iPXE,tag:X86PC,ltsp/undionly.kpxe
dhcp-boot=tag:!iPXE,tag:X86-64_EFI,ltsp/snponly.efi
dhcp-boot=tag:iPXE,ltsp/ltsp.ipxe

any help?

**Comment by alkisg** [comment]
It seems that dnsmasq has some issues setting next-server. Here's an example that worked for me.

1) Let's assume that LTSP server = 192.168.67.1. We only want the TFTP service of dnsmasq there, not real DHCP nor proxy DHCP. So, in the LTSP server, run:

```shell
ltsp dnsmasq -p0 -r0
```

2) Let's assume that real DHCP server = 192.168.67.10 (your infoblox). We only want the real DHCP service of dnsmasq there, not proxy DHCP nor TFTP. So, use this file:

```shell
# This file is part of LTSP, https://ltsp.org
# Copyright 2019 the LTSP team, see AUTHORS
# SPDX-License-Identifier: GPL-3.0-or-later

# Configure dnsmasq for LTSP
# Documentation=man:ltsp-dnsmasq(8)

# For additional local dnsmasq configuration like DNS blacklisting, it's
# recommended to use separate /etc/dnsmasq.d/local.conf files,
# so that they're not lost if you ever (re)run `ltsp dnsmasq`.

# port=0 disables the DNS service of dnsmasq
#port=0

# enable-tftp enables the TFTP service of dnsmasq
#enable-tftp

# FHS 2.3+ recommends /srv for tftp (debian #477109, LP #84615)
#tftp-root=/srv/tftp

# Log lots of extra information about DHCP transactions
#log-dhcp

# IP ranges to hand out, usually on the internal LTSP subnet of 2-NIC setups
dhcp-range=192.168.67.20,192.168.67.250,12h

# If another DHCP server is present on the network, a proxy range may be used
# instead. This makes dnsmasq provide boot information but not IP leases.


# Specify the DNS server. 0.0.0.0 means the machine running dnsmasq.
# DNS_SERVER in ltsp.conf is preferred as it reaches proxy DHCP clients.
dhcp-option=option:dns-server,0.0.0.0,8.8.8.8,208.67.222.222

# Set some tags to be able to separate client settings later on.
# "39" means "recent iPXE with menu support": http://ipxe.org/howto/dhcpd
dhcp-match=set:iPXE,175,39
dhcp-match=set:X86PC,option:client-arch,0
dhcp-match=set:X86-64_EFI,option:client-arch,7
# Due to rfc4578 errata, sometimes BC_EFI=9 is misused instead of X86-64_EFI=7:
dhcp-match=set:X86-64_EFI,option:client-arch,9
dhcp-mac=set:rpi,b8:27:eb:*:*:*
dhcp-mac=set:rpi,dc:a6:32:*:*:*

# In proxy DHCP mode, the server ONLY sends its IP and the following filename.
# Service types: man dnsmasq or https://tools.ietf.org/html/rfc4578#section-2.1
# PXE services in non proxy subnets sometimes break UEFI netboot, so tag:proxy.
#pxe-service=tag:proxy,tag:!iPXE,X86PC,"undionly.kpxe",ltsp/undionly.kpxe
#pxe-service=tag:proxy,tag:!iPXE,X86-64_EFI,"snponly.efi",ltsp/snponly.efi
#pxe-service=tag:proxy,tag:iPXE,X86PC,"ltsp.ipxe",ltsp/ltsp.ipxe
#pxe-service=tag:proxy,tag:iPXE,X86-64_EFI,"ltsp.ipxe",ltsp/ltsp.ipxe
#pxe-service=tag:rpi,X86PC,"Raspberry Pi Boot   ",unused

# Specify the boot filename for each tag, relative to tftp-root.
# If multiple lines with tags match, the last one is used.
# See: https://www.syslinux.org/wiki/index.php?title=PXELINUX#UEFI
dhcp-boot=tag:!iPXE,tag:X86PC,ltsp/undionly.kpxe,192.168.67.1,192.168.67.1
dhcp-boot=tag:!iPXE,tag:X86-64_EFI,ltsp/snponly.efi,192.168.67.1,192.168.67.1
dhcp-boot=tag:iPXE,ltsp/ltsp.ipxe,192.168.67.1,192.168.67.1

# Proxy DHCP clients don't receive any DHCP options like root-path.
# So we set root-path in the kernel cmdline from ltsp.ipxe.
#dhcp-option=option:root-path,ipxe-menu-item
```

The main thing to notice is that at the end of the `dhcp-boot` lines, I added the next-server option, the IP of the LTSP server.
While when I tried `dhcp-option=66,"192.168.67.1"`, it didn't work as expected.
You may ask in the dnsmasq mailing list for more details about dnsmasq configuration settings.

**Reply by HappyRizi** [reply]
Thank you Alkisg. It makes sans. specially the next-server option in the end of dhcp-boot line.
The command "ltsp config ..." does not work for me, it says  "Could not locate LTSP applet: config"
I am using "ltsp dnsmasq -p0 -r0" , is this "config" applet something I am missing?
I will try and let you know.

**Comment by alkisg** [comment]
Sorry `ltsp-config` was an old command of LTSP5. Multitasking too much does that to old people! :)
Thanks, I fixed it above.

**Reply by HappyRizi** [reply]
I have been giving up for a while and now I tried to look at the problem again. Here we need to send undionly.kpxe first and then ltsp.ipxe But the infoblox can only send one file for every pxe-request. DHCP options not working here. The dnsmasq does not work neither. My system do go to loop when I am sending the undionly.kpxe first (missing information in ltsp.ipxe). And when I send ltsp.ipxe so the client get no NIC-driver and no reaction here (missing undionly.kpxe). I fund that we need to break the loop. And I did it by an embedded script. I did a chain-loading (where ltsp.ipxe included in undionly.kpxe) then compiled a new undionly.kpxe file :
Create an embedded script file as below:
**=============(embed.ipxe)===========
#!ipxe
dhcp
chain tftp://${next-server}/ltsp/ltsp.ipxe || shell
==================================**

make undionly.kpxe EMBED=embed.ipxe 
and then upload the new undionly.kpxe to /srv/tftp/ltsp/
Boot client and it works fin!
I hope that it can help the other in the same situation.
